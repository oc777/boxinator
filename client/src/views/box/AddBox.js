import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { fetchCountries } from '../../store/actions/countryActions'
import { addBox } from '../../store/actions/boxActions'

class AddBox extends Component {
  state = {
    destination: 1,
    color: '255,255,255',
    hex: '#ffffff',
    nameReceiver: '',
    weight: ''
  }

  componentDidMount () {
    // fetch
    this.props.fetchCountries()
  }

  componentDidUpdate(prevProps) {
    if (this.props.boxData.success !== prevProps.boxData.success) {
        console.log('component updated')
        this.setState({
            destination: 1,
            color: '255,255,255',
            hex: '#ffffff',
            nameReceiver: '',
            weight: ''
        })
      }
  }

  handleChange = (event) => {
    const nam = event.target.name
    const val = event.target.value
    this.setState({ [nam]: val })
  }

  handleSubmit = (event) => {
    event.preventDefault()
    this.props.addBox(this.state)
  }

  handleColorChange = (event) => {
    const color = event.target.value
    const r = parseInt(color.substr(1, 2), 16)
    const g = parseInt(color.substr(3, 2), 16)
    const b = parseInt(color.substr(5, 2), 16)

    this.setState({
      hex: color,
      color: `${r},${g},${b}`
    })
  }

  getCountryList() {
    const data = this.props.countryData
    let list
    if (data.destination || []) {
      list = data.destination.map(item => {
        return (
          <option key={item.id} value={item.id}>{item.name}</option>
        )
      })
    }
    return list
  }

  render () {
    // console.log(this.props.boxData)
    //const data = this.props.data.box
    const list = this.getCountryList()

    return (
      <>
        <div className='addbox'>
          <h1>Add new Box</h1>
          <p>Fill in the form and click <code>Save</code></p>
        </div>

        <div className='info'>
          {this.props.boxData.pending && <p className='loading'>Sending...</p>}
          {this.props.boxData.success && <p className='loading'>Success!</p>}
          {this.props.boxData.error && <p className='error'>Something went terribly wrong here :(</p>}
        </div>

        <form className='add-box-form' onSubmit={this.handleSubmit}>
          <label>Name</label>
          <input name='nameReceiver' type='text'
          onChange={this.handleChange} value={this.state.nameReceiver}
          required />

          <label>Weight</label>
          <input name='weight' type='number' min='0' step='0.01'
          onChange={this.handleChange} value={this.state.weight}
          required />

          <label>Box colour</label>
          <div className='color-picker'>
            <label htmlFor='color'>- click to pick the colour -</label>
            <input
              name='color' id='color' type='color' value={this.state.hex}
              onChange={this.handleColorChange} required
            />
          </div>

          <label>Country</label>
          <select name='destination' value={this.state.country} onChange={this.handleChange}>
            {list}
          </select>

          <input type='submit' value='Save' />
        </form>
      </>
    )
  }
}

const mapStateToProps = state => ({
  boxData: state.boxReducer,
  countryData: state.countryReducer
})

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchCountries: fetchCountries,
  addBox: addBox
}, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddBox)
