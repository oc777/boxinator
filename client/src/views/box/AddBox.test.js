import { AddBox } from './AddBox'
import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

const defaultProps = {
  countryData: {
    destination:[
      {id: 1,name: "Sweden",multiplier: 1.3},
      {id: 2,name: "China",multiplier: 4}
    ]
  },
  boxData: {
    pending: false,
    errors: null,
    success: false
  },
  fetchCountries: jest.fn(),
  addBox: jest.fn()
}

describe('AddBox snapshot', () => {
  it('renders without crashing with required props', () => {
    const wrapper = renderer.create(<AddBox {...defaultProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})

describe('AddBox default state', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<AddBox {...defaultProps} />)
    // console.log(wrapper.debug())
    // console.log(wrapper.find('select').debug())
  })

  it('should render the form', () => {
    expect(wrapper.find('form').length).toBe(1)
    expect(wrapper.find('input').length).toBe(4)
  })

  it('should set correct value for color', () => {
    expect(wrapper.find('input[type="color"]').props().value).toBe('#ffffff')
  })

  it('should render select with correct options', () => {
    expect(wrapper.find('select').children().length).toBe(2)
  })
})

describe('AddBox handleChange', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<AddBox {...defaultProps} />)
  })

  it('should update state on change', () => {
    const name = wrapper.find('input[name="nameReceiver"]')
    const weight = wrapper.find('input[name="weight"]')
    const color = wrapper.find('input[name="color"]')
    const destination = wrapper.find('select[name="destination"]')

    name.simulate('change', { target: { name: 'nameReceiver', value: 'bob' } })
    expect(wrapper.state().nameReceiver).toMatch(/bob/)

    weight.simulate('change', { target: { name: 'weight', value: 1.2 } })
    expect(wrapper.state().weight).toBe(1.2)

    color.simulate('change', { target: { value: '#2a2a2a' }} )
    expect(wrapper.state().hex).toMatch(/2a2a2a/)

    destination.simulate('change', { target: { name: 'destination', value: 2 } })
    expect(wrapper.state().destination).toBe(2)
  })
})

describe('AddBox form validation', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallow(<AddBox {...defaultProps} />)
  })

  it('should assert error when input fields are empty', () => {
    const name = wrapper.find('input[name="nameReceiver"]')
    const weight = wrapper.find('input[name="weight"]')

    name.simulate('change', { target: { name: 'nameReceiver', value: '' } })
    expect(wrapper.state().nameError).toEqual('Fill in the name')
    weight.simulate('change', { target: { name: 'weight', value: '' } })
    expect(wrapper.state().weightError).toEqual('Weight must be over 0.1kg')
  })

  it('should assert error when color is blue', () => {
    const color = wrapper.find('input[name="color"]')

    color.simulate('change', { target: { value: '#03fce7' }})
    expect(wrapper.state().colorError).toEqual('All shades of blue are forbidden')
    
    color.simulate('change', { target: { value: '#3503fc' }})
    expect(wrapper.state().colorError).toEqual('All shades of blue are forbidden')
  })

  it('should validate when color is not blue', () => {
    const color = wrapper.find('input[name="color"]')

    color.simulate('change', { target: { value: '#03fc90' }})
    expect(wrapper.state().colorError).toEqual('')
    
    color.simulate('change', { target: { value: '#4503fc' }})
    expect(wrapper.state().colorError).toEqual('')
  })
})
