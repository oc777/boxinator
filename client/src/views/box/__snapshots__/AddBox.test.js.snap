// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBox snapshot renders without crashing with required props 1`] = `
Array [
  <div
    className="addbox"
  >
    <h1>
      Add new Box
    </h1>
    <p>
      Fill in the form and click 
      <code>
        Save
      </code>
    </p>
  </div>,
  <div
    className="info"
  />,
  <form
    className="add-box-form"
    onSubmit={[Function]}
  >
    <label>
      Name
    </label>
    <input
      name="nameReceiver"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <p
      className="input-err"
    >
      
    </p>
    <label>
      Weight
    </label>
    <input
      min="0"
      name="weight"
      onChange={[Function]}
      required={true}
      step="0.01"
      type="number"
      value=""
    />
    <p
      className="input-err"
    >
      
    </p>
    <label>
      Box colour
    </label>
    <div
      className="color-picker"
    >
      <label
        htmlFor="color"
      >
        - click to pick the colour -
      </label>
      <input
        id="color"
        name="color"
        onChange={[Function]}
        required={true}
        type="color"
        value="#ffffff"
      />
    </div>
    <p
      className="input-err"
    >
      
    </p>
    <label>
      Country
    </label>
    <select
      name="destination"
      onChange={[Function]}
    >
      <option
        value={1}
      >
        Sweden
      </option>
      <option
        value={2}
      >
        China
      </option>
    </select>
    <input
      type="submit"
      value="Save"
    />
  </form>,
]
`;
